### PR TITLE
[Draft][luci-interpreter] draft support for unroll gru

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This PR adds support for unroll gru model.

issue #9225 

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 1]
```diff
@@ -76,9 +76,8 @@ void Pack::configure()
     }
   }
 
-  if (t0->element_type() == DataType::S32 || t0->element_type() == DataType::U8 ||
-      t0->element_type() == DataType::S8 || t0->element_type() == DataType::S16 ||
-      t0->element_type() == DataType::S64)
+  if (t0->element_type() == DataType::U8 || t0->element_type() == DataType::S8 ||
+      t0->element_type() == DataType::S16)
```
Comment (BalyshevArtem) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:80]:
```c++
t0->zero_point()
```
fails due to `DataType::S32` and `DataType::S64` tensors don't have quantize params

File: compiler/luci/import/src/CircleReader.cpp [hunk 2]
```diff
@@ -264,10 +264,10 @@ void copy_tensor_attributes(const circle::Tensor *tensor, CircleNode *node)
...
   for (uint32_t r = 0; r < dims.size(); ++r)
   {
-    if (tensor_shape_signature.size() > 0 && tensor_shape_signature.at(r) == -1)
-      node->dim(r).unset();
-    else
-      node->dim(r).set(dims[r]);
+    // if (tensor_shape_signature.size() > 0 && tensor_shape_signature.at(r) == -1)
+    //  node->dim(r).unset();
+    // else
+    node->dim(r).set(dims[r]);
```
Comment (BalyshevArtem) [inline compiler/luci/import/src/CircleReader.cpp:270]: same as https://github.com/Samsung/ONE/pull/9253/files#r893606040

File: compiler/luci/import/src/Importer.cpp [hunk 3]
```diff
@@ -231,10 +231,10 @@ void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &r
...
     for (uint32_t r = 0; r < output_dims.size(); ++r)
     {
-      if (tensor_shape_signature.size() > 0 && tensor_shape_signature.at(r) == -1)
-        output_shape->dim(r).unset();
-      else
-        output_shape->dim(r).set(output_dims[r]);
+      // if (tensor_shape_signature.size() > 0 && tensor_shape_signature.at(r) == -1)
+      //  output_shape->dim(r).unset();
+      // else
+      output_shape->dim(r).set(output_dims[r]);
```
Comment (BalyshevArtem) [inline compiler/luci/import/src/Importer.cpp:237]: same as https://github.com/Samsung/ONE/pull/9253/files#r893606040

File: compiler/luci/import/src/Importer.cpp [hunk 4]
```diff
@@ -114,10 +114,10 @@ void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &r
...
     for (uint32_t r = 0; r < input_dims.size(); ++r)
     {
-      if (tensor_shape_signature.size() > 0 && tensor_shape_signature.at(r) == -1)
-        input_shape->dim(r).unset();
-      else
-        input_shape->dim(r).set(input_dims[r]);
+      // if (tensor_shape_signature.size() > 0 && tensor_shape_signature.at(r) == -1)
+      //   input_shape->dim(r).unset();
+      // else
+      input_shape->dim(r).set(input_dims[r]);
```
[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp:120]: This change is from #5682. @llFreetimell PTAL

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp:120]: @BalyshevArtem , can you please share the mode with `[0, some_value]` ?

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp:120]: maybe from https://github.com/Samsung/ONE/issues/9225#issuecomment-1147521316 ?

[the remark this answers is not in the document]
Comment (llFreetimell) [reply compiler/luci/import/src/Importer.cpp:120]: @BalyshevArtem 

> but if comment this (and also in CircleReader) then everything works fine and all tensor have right shapes.

### Current status

This problem maybe caused by unknown dimension with high probability.
For now, most of shape inference algorithm with unknown dimension is not supported.
(#5683 is created but not progressed because there are no actual needs for it)

I will explain current status this with [model_gru.circle.zip](https://github.com/Samsung/ONE/files/8845145/model_gru.circle.zip) in  https://github.com/Samsung/ONE/issues/9225#issuecomment-1147521316.
The model has following nodes.

- sequential_8/gru_8/transpose
  - shape : [28, 1, 28]
  - shape_signature : [28, -1, 28]
  - Thus this operation is represented as [28, None, 28] in python
- sequential_8/gru_8/unstack
  - shape : [1, 28]
  - shape_signature : [-1, 28]
  - Thus this operation is represented as [None, 28] in python

Please focus to `sequential_8/gru_8/gru_cell_8/MatMul;sequential_8/gru_8/gru_cell_8/BiasAdd` node.
This node is `CircleFullyConnected` and as a result following code is executed.

https://github.com/Samsung/ONE/blob/b297c4c4c34d3427a5b4269c7484cb5a66a26241/compiler/luci/service/src/CircleShapeInferenceRule.cpp#L727-L735

Since `input` of the node is `sequential_8/gru_8/unstack`, calculation is as below.
```cpp
 uint32_t input_size = 1; 

 // input_shape.dim(0),value() is 0 because first dimension is unset().
 // Thus, input_size is 0.
 for (uint32_t i = 0; i < input_shape.rank(); i++) 
 { 
   input_size = input_size * input_shape.dim(i).value(); 
 } 
 // Accordingly, batch_size is 0.
 const uint32_t batch_size = input_size / weights_shape.dim(1).value(); 
 out_shape.rank(2); 
 out_shape.dim(0) = batch_size; // Finally, 0 is inserted into output shape
 out_shape.dim(1) = weights_shape.dim(0); 
```

### Solution

For me, there are two solutions.

1. Implementation for each operation with considering unknown dimensions

    This way is standard solution, not workaround.
    Pros and cons are really obvious, large workloads but standard.

1. Add an option like `replaceUnknownDimensionAs1`

    This is workaround solution.
    If this option is turned up, all the unknown dimensions are set as 1.
    If this option is turned off, enable unknown dimensions with `unset()`.

[the remark this answers is not in the document]
Comment (BalyshevArtem) [reply compiler/luci/import/src/Importer.cpp:120]: @llFreetimell, thanks for the detailed explanation and solutions!
But I don't fully understand the first solution: `Implementation for each operation with considering unknown dimensions`.
Could you please tell what actually is the implementation for each operation, what actually should be done for this job, for example for CircleFullyConnected as in this model?
Thank you in advance :)

[the remark this answers is not in the document]
Comment (BalyshevArtem) [reply compiler/luci/import/src/Importer.cpp:120]: @seanshpark, 

> please share what problem you've encountered.

@llFreetimell in this comment https://github.com/Samsung/ONE/pull/9253#discussion_r894105961 described the problem, which I've encountered

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp:120]: I think it's `compiler/luci/service/include/luci/Service/CircleShapeInference.h` file.

[the remark this answers is not in the document]
Comment (llFreetimell) [reply compiler/luci/import/src/Importer.cpp:120]: @BalyshevArtem 

> Could you please tell what actually is the implementation for each operation

Briefly speaking, final goal of the first solution is "Fill in the tables in #5683"!

> what actually should be done for this job

Currently, almost all of shape inference algorithms are implemented with an assumption that every operation has no unknown dimension.
We should remove the assumption so that correct shape inference is done even when there are unknown dimensions :)

> for example for CircleFullyConnected

https://github.com/Samsung/ONE/blob/b297c4c4c34d3427a5b4269c7484cb5a66a26241/compiler/luci/service/src/CircleShapeInferenceRule.cpp#L727-L735

With above current implementation, following examples are wrongly inferenced.

- Example 1
  - input : [-1, 2]
  - weight : [3, 2]
  - Expected Output Shape : [-1, 3]
  - Actual Output Shape : [0, 3]
- Example 2
  - input : [4, 2]
  - weight : [3, -1] (FYI, weight can be non-constant but really rare case)
  - Expected Output Shape : [4, 3]
  - Actual Output Shape : Division by zero exception

So if we change the implementation as following codes, correct output shape can be inferenced for all cases.

```cpp
uint32_t batch_size = 1; 
bool has_unknown_dim = false;
for (uint32_t i = 0; i < input_shape.rank() - 1; i++) 
{
  if (input_shape.dim(i).known() == false)
  {
    has_unknown_dim = true;
    break;
  }
  batch_size *= input_shape.dim(i).value(); 
} 

out_shape.rank(2); 
if (has_unknown_dim)
  out_shape.dim(0).unset();
else
  out_shape.dim(0) = batch_size; 
out_shape.dim(1) = weights_shape.dim(0); 
```

This is "Upgrade shape inference algorithm with considering unknown dimension" in https://github.com/Samsung/ONE/issues/5683#issue-785034164 and the upgraded implementation will remove following `//`!

https://github.com/Samsung/ONE/blob/b297c4c4c34d3427a5b4269c7484cb5a66a26241/compiler/luci/service/include/luci/Service/CircleShapeInference.h#L77

(Please refer to https://github.com/Samsung/ONE/issues/5683#issue-785034164 to create PR for this)

This is why I said that "large worloads" is needed... there are a lot of operations and TC to implement 😅

[the remark this answers is not in the document]
Comment (BalyshevArtem) [reply compiler/luci/import/src/Importer.cpp]: @llFreetimell 
I chose the second (more easy solution from your solutions https://github.com/Samsung/ONE/pull/9253#discussion_r894105961), and added new option `replaceUnknownDimensionAs1`.

**Comment by BalyshevArtem** [comment]
@seanshpark, PTAL :)

File: compiler/luci/import/src/Importer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (binarman) [reply compiler/luci/import/src/Importer.cpp]: @llFreetimell 
I've looked at https://github.com/Samsung/ONE/issues/5501

Looks like intended behavior differs from current implementation:

```
   circle file     ---- import --->   luci/IR   ---- export --->   circle file
====================================================================================

   [1, 2, 3, 1]                                                    [1, 2, 3, 1]
     (shape)                        [?, 2, 3, 1]                      (shape)
                        ----->      (TensorShape)    ----->
  [-1, 2, 3, 1]                                                    [-1, 2, 3, 1]
(shape_signature)                                                (shape_signature)

====================================================================================

   [1, 2, 3, 1]                                                    [1, 2, 3, 1]
     (shape)                        [1, 2, 3, 1]                      (shape)
                        ----->      (TensorShape)    ----->
       []                                                                []
(shape_signature)                                                (shape_signature)
```

Could you clarify, maybe I misunderstood something?

[the remark this answers is not in the document]
Comment (binarman) [reply compiler/luci/import/src/Importer.cpp]: What do you think about following suggestion:
- dimension marked with -1 in shape_signature are considered as "variable"
- dimension value is stored in luci IR and considered as a default value for "variable" dimension

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp]: @binarman , if you are to modify `loco`, please talk with @parjong

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp]: I disagree adding `replaceUnknownDimensionAs1` as looks like hack.
I think adding `set_dynamic` is better.

[the remark this answers is not in the document]
Comment (llFreetimell) [reply compiler/luci/import/src/Importer.cpp]: @binarman @BalyshevArtem 

Sorry for late comment, I was absent last week because of other schedules T.T
**Following comments are only my own opinion, not mandatory :wink:**

---

> I think main problem here is not related with shape inference, but with model import.
> ...
> According to this example, imported should save information about original dimensions even if they are marked as unknown.

At first, you should know that dimension in `shape` is always 1 when dimension of `shape_signature` is -1.

To understand this, we should go to long time ago.
[When TFLite schema was "version 3", there are no  `shape_signature` and only `shape` is existed.](https://github.com/tensorflow/tensorflow/blob/a94cff5861a7560768faa798d7150cff35e9472a/tensorflow/lite/schema/schema_v3.fbs#L48-L65)

However, as concept of unknown dimension is introduced, `shape` is not enough to represent it.
With this reason, [TFLite schema "version 3a" introduced `shape_signature`.](https://github.com/tensorflow/tensorflow/blob/a94cff5861a7560768faa798d7150cff35e9472a/tensorflow/lite/schema/schema_v3a.fbs#L186-L213)

Nevertheless, `shape` is still alive to keep backward compatibility.
To resolve this issue, [TFLite converts `-1` in `shape_signature` to `1` in `shape`.](https://github.com/tensorflow/tensorflow/blob/2ea19cbb575d076b4f521d3603211c8316ad5f8f/tensorflow/compiler/mlir/lite/flatbuffer_export.cc#L844-L846)

In conclusion, `1` in `shape` is meaningless, which means that the value is not needed to be saved :)

> What do you think about new semantics of unknown dimensions?
> - dimension marked with -1 in shape_signature are considered as "variable"
> - dimension value is stored in luci IR and considered as a default value for "variable" dimension

As I said above, dimension value of `shape` is meaningless when the dimension of `shape_signature` is `-1`.
In addition, "unknown dimension has default dimension value" itself is weird because it means that "unknown is known"..!
That's why `unset` call sets dimension to 0.
(TMI : Dimension value can be 0 even when the dimension is known but it's not important now :smile: )

So I was against with introducing `set_dynamic(default_value)` function :)

> I chose the second (more easy solution from your solutions https://github.com/Samsung/ONE/pull/9253#discussion_r894105961), and added new option `replaceUnknownDimensionAs1`.

More precisely, my intention was "resetBatchSize` option like following example :)

```bash
$ circle2circle --reset_batch_size 2 model.circle model.opt.circle
```

In almost all cases, unknown dimensions are generated because of unknown batch size.
Since we use single batch in most of cases, the unknown dimension is usually reset to 1.
This issue is often detected when we work with new models :)

And if we introduce this feature as `luci::Pass`, it seems much safer thanmodifying `import()` function I think :D

File: compiler/luci/import/include/luci/Import/CircleReader.h [hunk 5]
```diff
@@ -48,7 +48,8 @@ std::unique_ptr<CircleQuantParam>
 luci_quantparam(const circle::QuantizationParameters *quantization);
 
 /// @brief Copy common tensor attributes such as name, type, etc. to node.
-void copy_tensor_attributes(const circle::Tensor *tensor, CircleNode *node);
+void copy_tensor_attributes(const circle::Tensor *tensor, CircleNode *node,
+                            bool replaceUnknownDimensionAs1 = false);
```
Comment (seanshpark) [inline compiler/luci/import/include/luci/Import/CircleReader.h:52]: please do not add `replaceUnknownDimensionAs1` this seems a hack.

Comment (seanshpark) [reply compiler/luci/import/include/luci/Import/CircleReader.h:52]: and we do not use initial value `= false` that may confuse API.

File: compiler/luci/import/src/Importer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (binarman) [reply compiler/luci/import/src/Importer.cpp]: @seanshpark 
I am convinced by @llFreetimell arguments about `replaceUnknownDimensionAs1` option )

**TL;DR version**
[TFLite converts -1 in shape_signature to 1 in shape.](https://github.com/tensorflow/tensorflow/blob/2ea19cbb575d076b4f521d3603211c8316ad5f8f/tensorflow/compiler/mlir/lite/flatbuffer_export.cc#L844-L846) is from TFLite converter:
``` c++
Optional<BufferOffset<tflite::Tensor>> Translator::BuildTensor(...)
  ...
  if (type.hasStaticShape()) {
    // shape fully defined
    ...
  } else if (inst && IsConst(inst)) {
    // handle special constant nodes with "dynamic" shape.
    // actual shape can be retrieved from operation attributes
    ...
  } else if (type.hasRank()) {
    ...
    
    shape.reserve(shape_ref.size());
    for (auto& dim : shape_ref) {
      shape.push_back(dim == -1 ? 1 : dim);
    }
    shape_signature = std::vector<int32_t>(shape_ref.begin(), shape_ref.end());
  }
  ...
}
```

If you wonder what is `type`, take a look: [mlir::TensorType](https://mlir.llvm.org/doxygen/classmlir_1_1TensorType.html)

What do you think?

P.S. Maybe `replaceUnknownDimensionAs1` is not the best name for the option, but it looks applicable.
Maybe we just need to add more detailed comment?

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp]: > ... arguments about replaceUnknownDimensionAs1 option )

Can be optional but I do not agree on current implementation.

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/src/Importer.cpp]: > Maybe replaceUnknownDimensionAs1 is not the best name for the option,

What I disagree is not about the name.

This implementation changed the API. I hope not to change for this and please do not use default value as current code uses `= false`.

Frankly speaking, I couldn't cat the code you wrote.
